### PR TITLE
Add flag to re-run tests multiple times

### PR DESCRIPTION
**Pull request by alexivanov**
### Changes

This PR adds an optional argument to `runAllTests`. As name suggests it runs tests up to N times, regardless of whether it has a difference to base on the head run.

This is useful for determining flake rate of a given session.

### CODE REVIEW

File: packages/cli/src/parallel-tests/__tests__/merge-test-results.spec.ts [hunk 1]
```diff
@@ -121,4 +123,16 @@ describe("mergeResults", () => {
       ])
     );
   });
+
+  it("returns a fail outcome if the original outcome was a pass and there are flakes", () => {
```
Comment (Que3216) [inline packages/cli/src/parallel-tests/__tests__/merge-test-results.spec.ts:127]: Isn't this basically identical to the "doesn't ignore diffs to screenshots which originally passed" test case, but with the `noDiff(0), diff(1)` order swapped? Wonder if we could just delete the "doesn't ignore diffs to screenshots which originally passed" test case.

**Review by Que3216 — COMMENTED** [review]
Added one comment, but haven't reviewed it all yet -- I'll review the rest tomorrow.

File: packages/cli/src/parallel-tests/__tests__/merge-test-results.spec.ts (hunk 1 above)
Comment (alexivanov) [reply packages/cli/src/parallel-tests/__tests__/merge-test-results.spec.ts:127]: oops that's true. will drop this test

File: packages/cli/src/parallel-tests/__tests__/parrallel-tests.handler.spec.ts [hunk 2]
```diff
@@ -201,6 +206,41 @@ describe("runAllTestsInParallel", () => {
       ),
     ]);
   });
+
+  it("retries N times if rerunTestsNTimes > 0", async () => {
+    let rerunNum = 0;
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    // const { baseTestRunId, ...testCaseWithoutBase } = testCase(0);
```
Comment (Que3216) [inline packages/cli/src/parallel-tests/__tests__/parrallel-tests.handler.spec.ts:213]:
```suggestion
```

File: packages/cli/src/parallel-tests/merge-test-results.ts [hunk 3]
```diff
@@ -111,6 +110,9 @@ export const mergeResults = ({
     result:
       currentResult.result === "fail" && noLongerHasFailures
         ? "flake"
+        : currentResult.result === "pass" &&
+          comparisonToHeadReplay.result === "fail"
+        ? "fail"
```
Comment (Que3216) [inline packages/cli/src/parallel-tests/merge-test-results.ts:115]: Should this be flake? (I'm assuming comparisonToHeadReplay is "fail" if any screenshot is different between the head and the retry)

Side note: I find the nested ? statements a bit hard to parse, so I'll sometimes split out into a function, or use if statements:

```
const screenshotDiffResultsByBaseReplayId: groupScreenshotDiffResults(
      newScreenshotDiffResults
    );
if (currentResult.result === "fail" && noLongerHasFailures) {
   return { ...currentResult, result: "flake", screenshotDiffResultsByBaseReplayId}
} else if (...) {
   ...
}
```

Comment (Que3216) [reply packages/cli/src/parallel-tests/merge-test-results.ts:115]: i.e. copy the logic from deflake-tests-handler:

```
  const result = hasNotableDifferences(
    Object.values(screenshotDiffResultsByBaseReplayId).flat()
  )
    ? "fail"
    : "pass";
```

Comment (Que3216) [reply packages/cli/src/parallel-tests/merge-test-results.ts:115]: Note: it may also make sense to refactor so that we no longer calculate the aggregated result. The backend could calculate it when it ingests the result or something. Otherwise it's a bit of a pain passing it around and mantaining it. I'm not sure we even use it anymore.

Comment (alexivanov) [reply packages/cli/src/parallel-tests/merge-test-results.ts]: https://github.com/alwaysmeticulous/meticulous-sdk/pull/335/commits/a3eb7bec3a2a6f2b4288ba0f5d1d0c19437d4d82 refactors the logic

**Review by Que3216 — APPROVED** [review]
